hugr-core/src/std_extensions/collections/borrow_array.rs

```diff
@@ -0,0 +1,688 @@
+//! A version of the standard fixed-length array extension that includes unsafe
+//! operations for borrowing and returning that may panic.
+
+use std::sync::{self, Arc};
+
+use delegate::delegate;
+use lazy_static::lazy_static;
+
+use crate::extension::{ExtensionId, SignatureError, TypeDef, TypeDefBound};
+use crate::ops::constant::{CustomConst, ValueName};
+use crate::types::type_param::{TypeArg, TypeParam};
+use crate::types::{CustomCheckFailure, Term, Type, TypeBound, TypeName};
+use crate::{Extension, Wire};
+use crate::{
+    builder::{BuildError, Dataflow},
+    extension::SignatureFunc,
+};
+use crate::{
+    extension::simple_op::{HasConcrete, HasDef, MakeExtensionOp, MakeOpDef, MakeRegisteredOp},
+    ops::ExtensionOp,
+};
+use crate::{
+    extension::{
+        OpDef,
+        prelude::usize_t,
+        resolution::{ExtensionResolutionError, WeakExtensionRegistry},
+        simple_op::{OpLoadError, try_from_name},
+    },
+    ops::OpName,
+    types::{FuncValueType, PolyFuncTypeRV},
+};
+
+use super::array::op_builder::GenericArrayOpBuilder;
+use super::array::{
+    Array, ArrayKind, FROM, GenericArrayClone, GenericArrayCloneDef, GenericArrayConvert,
+    GenericArrayConvertDef, GenericArrayDiscard, GenericArrayDiscardDef, GenericArrayOp,
+    GenericArrayOpDef, GenericArrayRepeat, GenericArrayRepeatDef, GenericArrayScan,
+    GenericArrayScanDef, GenericArrayValue, INTO,
+};
+
+/// Reported unique name of the borrow array type.
+pub const BORROW_ARRAY_TYPENAME: TypeName = TypeName::new_inline("borrow_array");
+/// Reported unique name of the borrow array value.
+pub const BORROW_ARRAY_VALUENAME: TypeName = TypeName::new_inline("borrow_array");
+/// Reported unique name of the extension
+pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("collections.borrow_arr");
+/// Extension version.
+pub const VERSION: semver::Version = semver::Version::new(0, 1, 1);
+
+/// A linear, unsafe, fixed-length collection of values.
+///
+/// Panic arrays are linear, even if their elements are copyable.
+#[derive(Clone, Copy, Debug, derive_more::Display, Eq, PartialEq, Default)]
+pub struct BorrowArray;
+
+impl ArrayKind for BorrowArray {
+    const EXTENSION_ID: ExtensionId = EXTENSION_ID;
+    const TYPE_NAME: TypeName = BORROW_ARRAY_TYPENAME;
+    const VALUE_NAME: ValueName = BORROW_ARRAY_VALUENAME;
+
+    fn extension() -> &'static Arc<Extension> {
+        &EXTENSION
+    }
+
+    fn type_def() -> &'static TypeDef {
+        EXTENSION.get_type(&BORROW_ARRAY_TYPENAME).unwrap()
+    }
+}
+
+/// Panic array operation definitions.
```

Comment:
```suggestion
/// Borrow array operation definitions.
```

same for the other docs